### DISCUSSION TITLE
how to shift the screen (i.e. to the left)?

### DISCUSSION THREAD

**Discussion by homonto** [Bodmer/TFT_eSPI#2551] [Q&A - General]
Hi @Bodmer 
I am not sure where to ask so feel free to delete the post please ;-)
I am working on the project with ILI9341, where big part of the screen is used to display life data (around 200 x 300px). The rest is not moving - just updating the values.
I have around 300px to work with (horizontal orientation) and I am printing the lines and pixels every second.
Of course, after 300 seconds I am ... lost - the graph is either pointing out of the screen or... 

**I see 2 options:**
1- clear the plot area and start from x=0 but that is not nice - history is lost
2- start shifting the plot area to the left: clean left side i.e by 50px, shift everything by 50px to the left having now extra 50px to accommodate next 50 seconds. And continue if needed - of course screen can accommodate only 300s with granularity 1s/1px.
This  option would be also great with shifting also up/down - not only left - as sometimes data incoming is going very down or very up and then the graph is either too close to the bottom or to the top.

I like the second option but I have NO B...Y idea how to touch it. I am looking for the examples but the only ones I found are text scrolling. But how about already printed graphs?

3- the last option (but probably even more difficult) would be: after 300 seconds passed, start compressing the graph and instead of plotting every 1s do it every 2s or so - but that would require much more as data is nowhere stored - just plotted - so maybe not the best option this one is.

Hopefully you or someone else would point me into the proper direction.

Thank you.

**Comment by Bodmer** [comment]
Have a look at the UTFT_Demo sketch, at around line 100 it draws a moving sine wave.

The old pixel y coordinate values are stored in an array so the position of each pixel is known, thus the old pixel can be erased and a new one draw. In your case the new one will be the next array value. i.e. erase array[n] pixel, draw array[n+1] pixel at that same x coord. When the last one is done, copy array [n+1] to array[n] and repeat for the next sample to be drawn.

**Comment by homonto** [comment]
thank you @Bodmer 
So in short: no library exists that does the magic (for lazy programmer like me) - it has to be written and then displayed again just with different values, right?
No problem. I will try ;-)

**Comment by homonto** [comment]
OK, I analysed it (by adding few delays here and there - otherwise it was too fast) - it looks neat: it starts from x=0 when end of the screen is reached, and draws again new screen but keeps deleting around 30px ahead, so that it looks like ... a snake game from old fashioned Nokia, right?
So that is option 4 that I never thought about ;-)
Thank you again @Bodmer

**Comment by arnolde** [comment]
Hi there,
I also would like to display a realtime moving graph on a 480x320 TFT (ili9488 spi).
I tried modifying the UTFT_DEMO sketch, here is what my loops look like:
```
  for (int i=0; i<10000; i++)
  {
    y=159+(sin(((i*0.7)*3.14)/180)*(90-(i / 100)));
    buf[480]=y;
    for (int x=0; x<480; x++) {
      tft.drawPixel(x,buf[x],TFT_BLACK);
      buf[x] = buf[x+1];
      tft.drawPixel(x,buf[x],TFT_WHITE);
    }
  }
```
But even with no delay() the moving sine wave is very slow (using ESP32 at 240Mhz and 40Mhz SPI speed), one cycle takes over 1 second to move across the screen. Are the 960 drawPixel actions THAT slow that each pixeldraw takes over a millisecond?
(I tried replacing the black pixels by a black rectangle before the inner loop but that flickers a lot, as expected, no good)

Is there any way to accelerate this? Is there really no "shift" or "scroll" function that simply and quickly moves the whole screen content over by 1 pixel?

**Comment by homonto** [comment]
sprite?

**Comment by arnolde** [comment]
To shift the whole screen?

**Comment by homonto** [comment]
why not?

**Comment by arnolde** [comment]
I'll try what this person demonstrates: https://www.youtube.com/watch?v=sRGXQg7028A
I'll still skeptical that a full-screen transfer operation is faster than 960 single pixels, but ok. And I'm still surprised that the ili9488 does not have a simple internal shift operation that would reduce the SPI communication to 2 small commands (shift and draw one pixel).

EDIT: Now I only get a blank grey screen with no changes or content.

```
void setup()
{
  tft.init();
  tft.setRotation(1);
  img.createSprite(480,320);
}

void loop()
{
  int y;
// Clear the screen and draw the frame
  img.fillScreen(TFT_BLACK);

  img.fillRect(0, 0, 480, 13, TFT_RED);

  img.fillRect(0, 305, 480, 320, TFT_GREY);
  img.setTextColor(TFT_BLACK,TFT_RED);

  img.drawCentreString("* TFT_eSPI *", CENTRE, 3, 1);
  img.setTextColor(TFT_YELLOW,TFT_GREY);
  img.drawCentreString("Adapted by Bodmer", CENTRE, 309,1);

  img.fillRect(1,15,478-1,304-15,TFT_BLACK);

  for (int i=0; i<10000; i++)
  {
    y=159+(sin(((i*0.7)*3.14)/180)*(90-(i / 100)));
    buf[480]=y;
    for (int x=0; x<480; x++) {
      img.drawPixel(x,buf[x],TFT_BLACK);
      buf[x] = buf[x+1];
      img.drawPixel(x,buf[x],TFT_WHITE);
    }
    img.pushSprite(0,0);
    delay(100);    
  }
  delay (500);
}
```

**Reply by Bodmer** [reply]
I suspect there is not sufficient free RAM to implement a full screen sprite so the sprite cannot be created.

Displays often have a hardware scroll capability,  this is used in the example here for an ILI9341 display.
https://github.com/Bodmer/TFT_eSPI/tree/master/examples/320%20x%20240/TFT_Terminal

Consult your driver data sheet for details. The area of the screen specified can only be scrolled up or down in portrait orientation and left or right in landscape orientation. A complication of that approach is the the coordinate datum also moves.

The ILI9488 is quite a slow screen to drive as it requires 3 colour bytes per pixel. Note that the screen interface protocol means14 bytes must be written to the screen to update 1 pixel, so using the SPI clock speed the pixel update rate can be calculated. 

The example here draws a waveform using lines and might give you some ideas.
https://github.com/Bodmer/ADC_DMA_FFT

**Reply by arnolde** [reply]
Hi Bodmer,
thanks for your input. I found a posting from someone who implemented this on the ili9488 in a language I don',t recognize: https://www.thebackshed.com/forum/ViewTopic_mobile.php?FID=16&TID=14436&PID=0&LastEntry=#179927
Here is also a video: https://youtu.be/NnuU_yy0B3Q
Looks very impressive! Any chance it could be implemented into your library? Unfortunately I dont have the time or priority (or knowhow) at the moment, but I think many people would find it very useful, since this ili9488 spi 320*480 display is extremely popular.

**Comment by arnolde** [comment]
@Bodmer Any chance the hardware scroll could (easily?) be implemented on the ILI9488 too, with the help of this info? It's already implemented for other chips, so maybe not too complicated?
